core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/If.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright 2017 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.gov.gchq.gaffer.operation.impl;
+
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.common.collect.Lists;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+import org.apache.commons.lang3.exception.CloneFailedException;
+
+import uk.gov.gchq.gaffer.operation.Operation;
+import uk.gov.gchq.gaffer.operation.OperationChain;
+import uk.gov.gchq.gaffer.operation.Operations;
+import uk.gov.gchq.gaffer.operation.io.InputOutput;
+import uk.gov.gchq.gaffer.operation.serialisation.TypeReferenceImpl;
+import uk.gov.gchq.gaffer.operation.util.Conditional;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * A {@code If} is an {@link Operation} which will execute one of two Operations,
+ * based on the result of testing an input Object against a provided {@link java.util.function.Predicate}.
+ *
+ * <p>This <code>Predicate</code> can also be configured with an {@link Conditional},
+ * which simply wraps an <code>Operation</code> and a <code> Predicate</code>.
+ * This enables pre-predicate transformation of the input,
+ * which allows properties other than the input object to be passed to the predicate,
+ * whilst preserving the initial input.</p>
+ *
+ * <p>As an example, this allows you to build an {@link Operation}
+ * which extracts a property from the input, passes it to the predicate,
+ * then the untouched original input is passed on to the <code>Operation</code> determined by the predicate test. </p>
+ *
+ * <p>A simple boolean, or anything that resolves to a boolean, can also be used to determine which <code>Operation</code> to execute. </p>
+ *
+ * @see If.Builder
+ */
+@JsonPropertyOrder(value = {"input", "condition", "predicate", "then", "otherwise", "options"}, alphabetic = true)
+public class If<I, O> implements InputOutput<I, O>,
+        Operations {
+
+    private I input;
+    private Map<String, String> options;
+    private Boolean condition;
+    private Operation then;
```

Comment:
Potentially, but I think this is too restrictive - this forces both <code>then</code> and <code>otherwise</code> to be of the same input/output type when in reality they may be very different - I've also just tried playing around with having them as a generic <code>InputOutput</code> but this also excludes operations such as <code>GetAllElements</code>, as well as <code>OperationChain</code> because they're only implementations of <code>Output</code>.